### PR TITLE
feat: less json requests, more faster

### PR DESCRIPTION
**Pull request by fengmk2**
using abbreviated meta

closes https://github.com/cnpm/npminstall/issues/218

---
This change is [Reviewable](https://reviewable.io/reviews/cnpm/npminstall/221)

### CODE REVIEW

File: lib/download/npm.js [hunk 1]
```diff
@@ -31,143 +29,95 @@ module.exports = function* (pkg, options) {
...
+    if (version) {
+      realPkgVersion = version;
+    } else if (range) {
+      realPkgVersion = semver.maxSatisfying(packageMeta.allVersions, range);
     }
   }
 
-  if (pkg.type === 'range' && RANGE_RE.test(pkg.spec)) {
-    const distTags = realPkg['dist-tags'];
-    if (distTags && (distTags.latest === realPkg.version || !semver.satisfies(distTags.latest, pkg.spec))) {
```
Comment (fengmk2) [inline lib/download/npm.js]: `semver.satisfies(distTags.latest, pkg.spec)` 这个逻辑漏了

**Comment by fengmk2** [comment]
等清明回来全部数据同步完成再合并发布

**Comment by fengmk2** [comment]
https://github.com/npm/registry/pull/156 还少了一个字段

**Comment by fengmk2** [comment]
先发个 beta 版本测试

```
npm dist-tag ls
beta: 3.0.0-beta.1
latest: 2.29.1
release-1: 1.15.0
```

**Comment by fengmk2** [comment]
cnpm 模块的 json 请求数对比

- before: All packages installed (181 packages installed from npm registry, used 8s, speed 846.4kB/s, json 202(163.9kB), tarball 6.61MB)
- after: All packages installed (181 packages installed from npm registry, used 9s, speed 817.97kB/s, json 177(337.49kB), tarball 6.61MB)

202 => 177 ~= 减少了 12%

---

egg 模块的 json 请求数对比

- before: All packages installed (202 packages installed from npm registry, used 7s, speed 343.3kB/s, json 237(192.25kB), tarball 2.29MB)
- after: All packages installed (202 packages installed from npm registry, used 6s, speed 434.21kB/s, json 184(252.49kB), tarball 2.29MB)

237 => 184 ~= 减少了 22%

**Comment by fengmk2** [comment]
https://github.com/cnpm/npminstall/blob/b340393db24a9a49f88dc90987dbc6415de4a413/lib/local_install.js#L414 recentlyUpdates 数据丢失

**Comment by fengmk2** [comment]
pick https://github.com/cnpm/npminstall/pull/227

File: test/bin.test.js [hunk 2]
```diff
@@ -15,7 +15,7 @@ describe('test/bin.test.js', () => {
   }
 
   beforeEach(cleanup);
-  afterEach(cleanup);
+  // afterEach(cleanup);
```
Comment (dead-horse) [inline test/bin.test.js]: 注释去掉吧

Comment (fengmk2) [inline test/bin.test.js]: 嗯，还有一个问题要解决了再继续改这个 pr。

**Comment by fengmk2** [comment]
@dead-horse 搞定了。

**Comment by fengmk2** [comment]
发个 3.0

**Pull request merged**